snapshot/lib/snapshot/simulator_launchers/simulator_launcher_xcode_8.rb

```diff
@@ -0,0 +1,107 @@
+              locale = language[1]
+              language = language[0]
+            end
+            results[device] ||= {}
+
+            current_run = device_index * launcher_config.languages.count + language_index + 1
+            number_of_runs = launcher_config.languages.count * launcher_config.devices.count
+            UI.message("snapshot run #{current_run} of #{number_of_runs}")
+            results[device][language] = run_for_device_and_language(language, locale, device, launch_args)
+            copy_simulator_logs([device], language, locale, launch_args)
+          end
+        end
+      end
+      results
+    end
+
+    # This is its own method so that it can re-try if the tests fail randomly
+    # @return true/false depending on if the tests succeeded
+    def run_for_device_and_language(language, locale, device, launch_arguments, retries = 0)
+      return launch_one_at_a_time(language, locale, device, launch_arguments)
+    rescue => ex
+      UI.error ex.to_s # show the reason for failure to the user, but still maybe retry
+
+      if retries < launcher_config.number_of_retries
+        UI.important "Tests failed, re-trying #{retries + 1} out of #{launcher_config.number_of_retries + 1} times"
+        run_for_device_and_language(language, locale, device, launch_arguments, retries + 1)
+      else
+        UI.error "Backtrace:\n\t#{ex.backtrace.join("\n\t")}" if FastlaneCore::Globals.verbose?
+        self.collected_errors << ex
+        raise ex if launcher_config.stop_after_first_error
+        return false # for the results
+      end
+    end
+
+    # Returns true if it succeeded
+    def launch_one_at_a_time(language, locale, device_type, launch_arguments)
+      prepare_for_launch(language, locale, launch_arguments)
+
+      add_media([device_type], :photo, launcher_config.add_photos) if launcher_config.add_photos
+      add_media([device_type], :video, launcher_config.add_videos) if launcher_config.add_videos
+
+      open_simulator_for_device(device_type)
+
+      command = TestCommandGeneratorXcode8.generate(device_type: device_type, language: language, locale: locale)
+
+      if locale
+        UI.header("#{device_type} - #{language} (#{locale})")
+      else
+        UI.header("#{device_type} - #{language}")
+      end
+
+      prefix_hash = [
+        {
+          prefix: "Running Tests: ",
+          block: proc do |value|
+            value.include?("Touching")
+          end
+        }
+      ]
+
+      FastlaneCore::CommandExecutor.execute(command: command,
+                                          print_all: true,
+                                      print_command: true,
+                                             prefix: prefix_hash,
+                                            loading: "Loading...",
+                                              error: proc do |output, return_code|
+                                                ErrorHandler.handle_test_error(output, return_code)
```

Comment:
I know this probably like this before, but it might be worth moving this into its own function for better readability 👍